raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,172 @@
+
+            updated_state_changes.append(StateChangeRecord(
+                state_change_identifier=state_change.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+    storage.update_state_changes(updated_state_changes)
+
+
+def _add_blockhash_to_events(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to all ContractSendXXX events"""
+    events = storage.get_all_event_records()
+    updated_events = []
+    for event in events:
+        data = json.loads(event.data)
+        if 'raiden.transfer.events.ContractSend' in data['_type']:
+            assert 'triggered_by_block_hash' not in data, 'v18 events cant contain blockhash'
+            # Get the state_change that triggered the event and if it has
+            # a block number get its hash. If not fall back to latest.
+            matched_state_changes = storage.get_statechanges_by_identifier(
+                from_identifier=event.state_change_identifier,
+                to_identifier=event.state_change_identifier,
+            )
+            result_length = len(matched_state_changes)
+            msg = 'multiple state changes should not exist for the same identifier'
+            assert result_length == 0 or result_length == 1, msg
+            block_hash = None
+            if result_length == 1:
+                statechange_data = json.loads(matched_state_changes[0])
+                if 'block_hash' in statechange_data:
+                    block_hash = statechange_data['block_hash']
+                elif 'block_number' in statechange_data:
+                    block_number = int(statechange_data['block_number'])
+                    block_hash = cache.get(block_number)
+
+            # else fallback to just using the latest blockhash
+            if not block_hash:
+                block_hash = cache.web3.eth.getBlock('latest')['hash']
+                block_hash = block_hash.hex()
+
+            data['triggered_by_block_hash'] = block_hash
+
+            updated_events.append(EventRecord(
+                event_identifier=event.event_identifier,
+                state_change_identifier=event.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+    storage.update_events(updated_events)
+
+
+def _transform_snapshot(
+        raw_snapshot: Dict[Any, Any],
+        storage: SQLiteStorage,
+        cache: BlockHashCache,
+) -> str:
+    """Upgrades a single snapshot by adding the blockhash to it and to any pending transactions"""
+    snapshot = json.loads(raw_snapshot)
+    block_number = int(snapshot['block_number'])
+    snapshot['block_hash'] = cache.get(block_number)
+
+    all_events = storage.get_all_event_records()
+    pending_transactions = snapshot['pending_transactions']
+
+    new_pending_transactions = []
+    for transaction in pending_transactions:
+        found_blockhash = None
+        transaction_data = json.loads(transaction)
```

Comment:
So you're getting the `pending_transactions` from `snapshot` which was already json-parsed. I don't think you need to parse it again. Am i missing something here?